lib/include/openamp/rpmsg.h

```diff
@@ -147,8 +168,17 @@ int rpmsg_send_offchannel_raw(struct rpmsg_endpoint *ept, uint32_t src,
 static inline int rpmsg_send(struct rpmsg_endpoint *ept, const void *data,
 			     int len)
 {
-	return rpmsg_send_offchannel_raw(ept, ept->addr, ept->dest_addr, data,
-					 len, true);
+	int tc = 0;
+
+	for (; tc < RPMSG_TICK_COUNT; tc += RPMSG_TICKS_PER_INTERVAL) {
+		if (is_rpmsg_ept_ready(ept))
+			return rpmsg_send_offchannel_raw(ept, ept->addr,
+							 ept->dest_addr,
+							 data, len, true);
+		metal_sleep_usec(RPMSG_TICKS_PER_INTERVAL);
+	}
+
```

Comment:
> This patch add an additional check on each send.
> Then what's happen if case of timeout in your application? you have another loop to try?

No, this loop will wait for 15 seconds, I don't think that more wait can help here.

> Could you explain the rational of this?

The patch is designed work with: https://github.com/OpenAMP/open-amp/pull/160
the caller can do:
```
rpmsg_create_ept
rpmsg_send
```
just like the socket API.

> i don't see here a real need except for NS_ACK, and only to try to simplify the application code.

It also work with your favor solution(the dummy server message). Sometime, it's very natual that the client need send the initial message to server just after the channel get created. Without this patch, you have to wait the sever send the first dummy back which make the code more complex than needed. And become impossible if the two phase handshake done inside the rpmsg except we provide the new callback to user.

> And you mention in previous commit: rpmsg_send_offchannel_raw already checks the destination address

Yes, but rpmsg_send_offchannel_raw return failure without the retry logic.

Reply:
The desgin is following the same principle inside rpmsg_virtio_send_offchannel_raw loop:
```
static int rpmsg_virtio_send_offchannel_raw(struct rpmsg_device *rdev,
					    uint32_t src, uint32_t dst,
					    const void *data,
					    int size, int wait)
{
	...
	if (wait)
		tick_count = RPMSG_TICK_COUNT / RPMSG_TICKS_PER_INTERVAL;
	else
		tick_count = 0;

	while (1) {
		...
		buffer = rpmsg_virtio_get_tx_buffer(rvdev, &buff_len, &idx);
		metal_mutex_release(&rdev->lock);
		if (buffer || !tick_count)
			break;
		if (avail_size != 0)
			return RPMSG_ERR_BUFF_SIZE;
		metal_sleep_usec(RPMSG_TICKS_PER_INTERVAL);
		tick_count--;
	}
```
Does this mean that we shold drop the above loop too? From the design, rpmsg_send should block and wait the resource available, only rpmsg_try_send can return fail directly wihtout wait.

Reply:
> The desgin is following the same principle inside rpmsg_virtio_send_offchannel_raw loop:
> 
> ```
> static int rpmsg_virtio_send_offchannel_raw(struct rpmsg_device *rdev,
> 					    uint32_t src, uint32_t dst,
> 					    const void *data,
> 					    int size, int wait)
> {
> 	...
> 	if (wait)
> 		tick_count = RPMSG_TICK_COUNT / RPMSG_TICKS_PER_INTERVAL;
> 	else
> 		tick_count = 0;
> 
> 	while (1) {
> 		...
> 		buffer = rpmsg_virtio_get_tx_buffer(rvdev, &buff_len, &idx);
> 		metal_mutex_release(&rdev->lock);
> 		if (buffer || !tick_count)
> 			break;
> 		if (avail_size != 0)
> 			return RPMSG_ERR_BUFF_SIZE;
> 		metal_sleep_usec(RPMSG_TICKS_PER_INTERVAL);
> 		tick_count--;
> 	}
> ```
> 
> Does this mean that we shold drop the above loop too? From the design, rpmsg_send should block and wait the resource available, only rpmsg_try_send can return fail directly wihtout wait.

No, the loop has to be kept here, blocking fro rpmsg_send, not blocking for rpmsg_try_send ( same behavior than Linux API).

Reply:
If so, why you consider my patch useless? It also handle the case where the resouce unavailable.

Reply:
Simply  because you have the API to do this in application, so i don't see a strong reason to integrate it in rpmsg_send.

Reply:
If so, we just need provide rpmsg_try_send I don't see any reason to add rpmsg_send which is very easy to implement on top of rpmsg_try_send. As I said before, it's a common issue all people will hit immediately after he/she create a rpmsg endpoint proactively. The library need provide a solution to help people mitigate the problem, instead let user find the touch bug and repeat the code again and again.

Reply:
> If so, we just need provide rpmsg_try_send I don't see any reason to add rpmsg_send which is very easy to implement on top of rpmsg_try_send.

rpmsg_try_send returns an error if no TX buffer available, rpmsg_send is blocked waiting a TX_buffer. No relation chip with your usecase

> As I said before, it's a common issue all people will hit immediately after he/she create a rpmsg endpoint proactively. The library need provide a solution to help people mitigate the problem, instead let user find the touch bug and repeat the code again and again.

As rpmsg_send returns an error if remote address is not defined it is more a bug in application that does not treat error and /or does not use is_rpmsg_ept_ready as show in examples.
So the issue here seems more the use of the API than the API itself. 
If user code is inspirated from the examples they should use is_rpmsg_ept_ready. 
But If you have implemented the NS ack for your customer you probably also increases this bad usage, by masking the rpmsg current protocol.

the solution to help user seems more the documentation or example.

Reply:
If you search examples folder, all client side has this code:
```
	/* Create RPMsg endpoint */
	ret = rpmsg_create_ept(&lept, rdev, RPMSG_SERVICE_NAME,
			       RPMSG_ADDR_ANY, RPMSG_ADDR_ANY,
			       rpmsg_endpoint_cb, rpmsg_service_unbind);
	if (ret) {
		LPERROR("Failed to create RPMsg endpoint.\r\n");
		return ret;
	}

	while (!is_rpmsg_ept_ready(&lept))
		platform_poll(priv);
```
As OpenAMP maintainer, do you have any responsbility to improve this situation in the next release? I don't think the code like this can survive in Linux kernel. Actually from my point, this patch is just the first step, the next is:
1.Ensure all system under libmetal/lib/lib implement conditon
2.Replace the poll loop with the event trigger just like Linux kernel
But anyway, I don't make any suggestion like this anymore since I don't want to waste my time to dissuss how to improve OpenAMP.